R/rnet_junctions.R

```diff
@@ -0,0 +1,30 @@
+#' Get points where 3 or more linestrings meet (junctions on road networks)
+#'
+#' @inheritParams rnet_breakup_vertices
+#' @export
+#' @examples
+#' has_sfheaders <- requireNamespace("sfheaders", quietly = TRUE)
+#' if(has_sfheaders) {
+#' rnet <- rnet_roundabout
+#' j <- rnet_junction(rnet)
+#' plot(rnet$geometry)
+#' plot(j, add = TRUE)
+#' }
+rnet_junction <- function(rnet) {
```

Comment:
TBH, I'm not sure why you called this function `rnet_junction` since (I think) it simply returns the boundary points of each LINESTRING. I would have called it `rnet_boundary_points` (or something like that). For example: 

``` r
# packages
library(stplanr)

rnet <- rnet_roundabout
j <- rnet_junction(rnet)

par(mar = rep(0, 4))
plot(rnet_roundabout$geometry)
plot(j, add = TRUE, pch = 16)
plot(j$geometry[1], add = TRUE, pch = 16, col = "red")
```

![](https://i.imgur.com/zgyT4Zc.png)

<sup>Created on 2020-11-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Clearly, the red dot is not a junction point, but it's the (unique) boundary point of the LINESTRING that defines the roundabout. Moreover, the output should be more or less the same as `line2points` (without the `id` column).

Reply:
Good point about the name @agila5, I was thinking of another function. In fact I think this could feed-into a more efficient implementation of the `junction()` function demonstrated below. Will updated the name :+1: 

``` r
# Aim: test 'rnet_junction()'
    remotes::install_github("ropensci/stplanr", "boundary-points")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'stplanr' from a github remote, the SHA1 (8eafeb87) has not changed since last install.
#>   Use `force = TRUE` to force installation
    remotes::install_github("saferactive/trafficalmr")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'trafficalmr' from a github remote, the SHA1 (1388bc79) has not changed since last install.
#>   Use `force = TRUE` to force installation
    rnet = trafficalmr::tc_data_osm
    p1 = stplanr::rnet_junction(rnet)
    p2 = trafficalmr::osm_get_junctions(rnet)
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 7.0.0
#> Splitting rnet object at the shared boundary points.
#> Splitting rnet object at the shared internal points.
    p3 = trafficalmr::osm_get_junctions(rnet, method = "duplicates")
#> Warning in st_cast.sf(x, "POINT"): repeating attributes for all sub-geometries
#> for which they may not be constant
    
    plot(rnet$geometry, col = "grey")
    plot(p1, add = T)
    plot(p2, add = T, col = "blue", cex = 1.5)
    plot(p2, add = T, col = "red", cex = 0.7)
```

![](https://i.imgur.com/4HlW5bB.png)

<sup>Created on 2020-11-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
Update: it seems that they are quite different. There are junction that are not boundary points. Could identifying these also be important for routing @agila5 ? I guess so but not sure. Reprex to illustrate what I mean:

``` r
# Aim: test 'rnet_junction()'
remotes::install_github("ropensci/stplanr", "boundary-points")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'stplanr' from a github remote, the SHA1 (8eafeb87) has not changed since last install.
#>   Use `force = TRUE` to force installation
remotes::install_github("saferactive/trafficalmr")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'trafficalmr' from a github remote, the SHA1 (1388bc79) has not changed since last install.
#>   Use `force = TRUE` to force installation
rnet = trafficalmr::tc_data_osm
p1 = stplanr::rnet_junction(rnet)
p2 = trafficalmr::osm_get_junctions(rnet)
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 7.0.0
#> Splitting rnet object at the shared boundary points.
#> Splitting rnet object at the shared internal points.
p3 = trafficalmr::osm_get_junctions(rnet, method = "duplicates")
#> Warning in st_cast.sf(x, "POINT"): repeating attributes for all sub-geometries
#> for which they may not be constant

nrow(rnet)
#> [1] 499
nrow(p1)
#> [1] 998
length(p2)
#> [1] 339
length(p3)
#> [1] 523


plot(rnet$geometry, col = "grey")
plot(p1, add = T)
plot(p2, add = T, col = "blue", cex = 1.5)
plot(p2, add = T, col = "red", cex = 0.7)
```

![](https://i.imgur.com/9HpNgFi.png)

``` r
summary(sf::st_geometry(p1) %in% sf::st_geometry(p2))
#>    Mode   FALSE    TRUE 
#> logical     493     505
summary(p2_in_p1 <- sf::st_geometry(p2) %in% sf::st_geometry(p1))
#>    Mode   FALSE    TRUE 
#> logical      50     289
library(tmap)
tmap_mode("view")
#> tmap mode set to interactive viewing
qtm(p1, dots.col = "grey", dots.size = 0.2) +
  qtm(p2[!p2_in_p1], dots.col = "blue", dots.size = 0.1) 
```

![](https://i.imgur.com/7rWqHYW.png)

<sup>Created on 2020-11-21 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
Here's an example of a 'junction point' that is not a 'boundary point' illustrating that these are quite different concepts:

![image](https://user-images.githubusercontent.com/1825120/99887217-6dfdb180-2c3a-11eb-9792-464edb227628.png)

Reply:
I think that all junction-points should be boundary-points after running `rnet_breakup_vertices` (or the new replacement with data.table). So (at least in my mind) the workflow is 1) get OSM data; 2) fix OSM data using rnet_breakup (or the new replacement); 3) identify junction points as the boundary points that are duplicated at least three times. The problem is that I cannot really run the previous reprex since you replaced `rnet_junction()`. The replacement is `rnet_boundary_points()`?

Reply:
>  The problem is that I cannot really run the previous reprex since you replaced `rnet_junction()`. The replacement is `rnet_boundary_points()`?

Correct

Reply:
Ok, I think I get it now. The problem with the previous reprex is that the default method of `osm_get_junctions()` modifies the input `rnet` using `rnet_breakup_vertices` (see [here](https://github.com/saferactive/trafficalmr/blob/ea6d36fbb41a14cfca4b8e18a669f20c42be25ad/R/osm_cleaning_functions.R#L163)), so the input LINESTRINGS are splitted at the shared interior points and the points shared between the boundaries and the interiors. On the other side, you forgot to apply `rnet_breakup_vertices` before running `rnet_boundary_points`. For example: 

``` r
# dev packages
remotes::install_github("ropensci/stplanr", "boundary-points")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'stplanr' from a github remote, the SHA1 (de5ef4cf) has not changed since last install.
#>   Use `force = TRUE` to force installation
remotes::install_github("saferactive/trafficalmr")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'trafficalmr' from a github remote, the SHA1 (1388bc79) has not changed since last install.
#>   Use `force = TRUE` to force installation

# packages
library(sf)
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1

# data
rnet = trafficalmr::tc_data_osm 

# wrong approach
p1 = st_geometry(stplanr::rnet_boundary_points(rnet))
p2 = trafficalmr::osm_get_junctions(rnet)
#> Splitting rnet object at the shared boundary points.
#> Splitting rnet object at the shared internal points.

summary(p2 %in% p1)
#>    Mode   FALSE    TRUE 
#> logical      50     289

# right approach
p1 = st_geometry(stplanr::rnet_boundary_points(rnet %>% stplanr::rnet_breakup_vertices()))
#> Splitting rnet object at the shared boundary points.
#> Splitting rnet object at the shared internal points.
p2 = trafficalmr::osm_get_junctions(rnet)
#> Splitting rnet object at the shared boundary points.
#> Splitting rnet object at the shared internal points.

summary(p2 %in% p1)
#>    Mode    TRUE 
#> logical     339
```

<sup>Created on 2020-11-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

I think that solves the problem.